Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -107,98 +125,141 @@
                                  1 -1 -1  1  1 -1 -1  1;
                                  1  1 -1 -1  1  1 -1 -1;
                                  1 -1  1 -1  1 -1  1 -1];
-		end    
-    end
+		end
+	end
     
-    if strcmp(xQ.Vendor,'Philips')
-        xQ.TimeEncodedMatrix = xQ.TimeEncodedMatrix * -1;
+    switch xQ.Vendor
+        case 'Siemens'
+            % This is where this code was initially based on, so for
+            % Siemens we don't do anything specific
+        case 'Philips'
+            xQ.TimeEncodedMatrix = xQ.TimeEncodedMatrix * -1;
+            % Philips uses an inverse matrix compared to Siemens
+        case 'GE'
+            warning('Time encoded ASL data detected for GE, but GE usually does the decoding in K-space on the scanner');
+        otherwise
+            warning(['Time encoded ASL data not yet tested for vendor ' xQ.vendor]);
     end
     
 end
 
-% Load time-series nifti
+
+%% 2. Load time-series nifti
 imEncoded = xASL_io_Nifti2Im(imPath); 
 
-% Calculate specific data sizes
-nDecodedTI = xQ.TimeEncodedMatrixSize-1;                   % Number of TIs is always MatrixSize -1
-nDecodedVolumes = xQ.NumberEchoTimes * nDecodedTI;
+% Prepare the general sequence parameters
+nEncodedVolumes = size(imEncoded, 4);     % Size of the raw data
+
+% Check that input quantification parameters are vectors with an equal length to the number of volumes
+if length(xQ.LabelingDuration)>1 && length(xQ.LabelingDuration) ~= nEncodedVolumes
+	error('Length of LabelingDuration vector does not match the number of volumes.');
+end
+
+if length(xQ.Initial_PLD)>1 && length(xQ.Initial_PLD) ~= nEncodedVolumes
+	error('Length of Initial_PLD vector does not match the number of volumes.');
+end
+
+% Prepare the parameters also for interleaved reordering
+% We thus calculate the size of each Hadamard block as the number of Hadamard phases and TEs
+nVolumesPerRepetition = xQ.TimeEncodedMatrixSize * xQ.NumberEchoTimes;
```

Comment:
This should be moved down, outside of the section that does the parameter checks.

Reply:
OK.